src/input.rs

```diff
@@ -0,0 +1,172 @@
+            mouse_pressed: HashSet::new(),
+            mouse_pos: [0.0; 2].into(),
+        };
+        let diff = InputDiff {
+            time_delta: 0.0,
+            keys_hit: Vec::new(),
+            mouse_moves: Vec::new(),
+            mouse_hit: Vec::new(),
+        };
+        Input(state, diff)
+    }
+
+    pub fn reset(&mut self) {
+        let now = time::Instant::now();
+        let dt = now - self.0.time_moment;
+        self.0.time_moment = now;
+        self.1.time_delta = dt.as_secs() as f32 + 1e-9 * dt.subsec_nanos() as f32;
+        self.1.keys_hit.clear();
+        self.1.mouse_moves.clear();
+        self.1.mouse_hit.clear();
+    }
+
+    pub fn time(&self) -> Timer {
+        Timer {
+            start: self.0.time_moment
+        }
+    }
+
+    pub fn get_time_delta(&self) -> f32 {
+        self.1.time_delta
+    }
+
+    pub fn get_mouse_pos(&self) -> mint::Point2<f32> {
+        self.0.mouse_pos
+    }
+
+    pub fn keyboard_input(&mut self, state: ElementState, key: Key) {
+        match state {
+            ElementState::Pressed => {
+                self.0.keys_pressed.insert(key);
+                self.1.keys_hit.push(key);
+            }
+            ElementState::Released => {
+                self.0.keys_pressed.remove(&key);
+            }
+        }
+    }
+
+    pub fn mouse_moved(&mut self, pos: mint::Point2<f32>) {
+        self.1.mouse_moves.push(mint::Vector2 {
+            x: pos.x - self.0.mouse_pos.x,
+            y: pos.y - self.0.mouse_pos.y,
+        });
+        self.0.mouse_pos = pos;
+    }
+}
+
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Timer {
+    start: time::Instant,
+    //focus_only: bool,
+}
+
+impl Timer {
+    pub fn get(&self, input: &Input) -> TimerDuration {
+        let dt = input.0.time_moment - self.start;
+        dt.as_secs() as f32 + 1e-9 * dt.subsec_nanos() as f32
+    }
+}
+
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Button {
+    Key(Key),
+    Mouse(MouseButton),
+}
+
+pub const KEY_ESCAPE: Button = Button::Key(Key::Escape);
+pub const KEY_SPACE: Button = Button::Key(Key::Space);
+
+impl Button {
+    pub fn count_hits(&self, input: &Input) -> u8 {
+        match *self {
+            Button::Key(button) => input.1.keys_hit.iter().filter(|&&key| key == button).take(0xFF).count() as u8,
+            Button::Mouse(button) => input.1.mouse_hit.iter().filter(|&&key| key == button).take(0xFF).count() as u8,
+        }
+    }
+
+    pub fn is_hit(&self, input: &Input) -> bool {
+        match *self {
+            Button::Key(button) => input.0.keys_pressed.contains(&button),
+            Button::Mouse(button) => input.0.mouse_pressed.contains(&button),
+        }
+    }
+}
+
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct KeyAxis {
+    pub neg: Key,
+    pub pos: Key,
+}
+
+pub const AXIS_LEFT_RIGHT: KeyAxis = KeyAxis{ neg: Key::Left, pos: Key::Right };
+pub const AXIS_DOWN_UP: KeyAxis = KeyAxis{ neg: Key::Down, pos: Key::Up };
```

Comment:
The axis go from DOWN (negative) to UP (positive), hence the name `AXIS_DOWN_UP`, which is consistent with `AXIS_LEFT_RIGHT`.